consensus/src/dag/reconstruction/mod.rs

```diff
@@ -0,0 +1,368 @@
+    /// We need the explicit list of parents for the unit identified by the hash.
+    /// This should only happen in the presence of forks, when optimistic reconstruction failed.
+    ParentsOf(H::Hash),
+}
+
+/// The result of a reconstruction attempt. Might contain multiple reconstructed units,
+/// as well as requests for some data that is needed for further reconstruction.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ReconstructionResult<U: Unit> {
+    reconstructed_units: Vec<ReconstructedUnit<U>>,
+    requests: Vec<Request<U::Hasher>>,
+}
+
+impl<U: Unit> ReconstructionResult<U> {
+    fn new(
+        reconstructed_units: Vec<ReconstructedUnit<U>>,
+        requests: Vec<Request<U::Hasher>>,
+    ) -> Self {
+        ReconstructionResult {
+            reconstructed_units,
+            requests,
+        }
+    }
+
+    fn empty() -> Self {
+        ReconstructionResult::new(Vec::new(), Vec::new())
+    }
+
+    fn reconstructed(unit: ReconstructedUnit<U>) -> Self {
+        ReconstructionResult {
+            reconstructed_units: vec![unit],
+            requests: Vec::new(),
+        }
+    }
+
+    fn request(request: Request<U::Hasher>) -> Self {
+        ReconstructionResult {
+            reconstructed_units: Vec::new(),
+            requests: vec![request],
+        }
+    }
+
+    fn add_unit(&mut self, unit: ReconstructedUnit<U>) {
+        self.reconstructed_units.push(unit);
+    }
+
+    fn add_request(&mut self, request: Request<U::Hasher>) {
+        self.requests.push(request);
+    }
+
+    fn accumulate(&mut self, other: ReconstructionResult<U>) {
+        let ReconstructionResult {
+            mut reconstructed_units,
+            mut requests,
+        } = other;
+        self.reconstructed_units.append(&mut reconstructed_units);
+        self.requests.append(&mut requests);
+    }
+}
+
+impl<U: Unit> From<ReconstructionResult<U>>
+    for (Vec<ReconstructedUnit<U>>, Vec<Request<U::Hasher>>)
+{
+    fn from(result: ReconstructionResult<U>) -> Self {
+        let ReconstructionResult {
+            reconstructed_units,
+            requests,
+        } = result;
+        (reconstructed_units, requests)
+    }
+}
+
+/// The reconstruction of the structure of the Dag.
+/// When passed units containing control hashes, and responses to requests it produces,
+/// it eventually outputs versions with explicit parents in an order conforming to the Dag order.
+pub struct Reconstruction<U: Unit> {
+    parents: ParentReconstruction<U>,
+    dag: Dag<U>,
+}
+
+impl<U: Unit> Reconstruction<U> {
+    /// Create a new reconstruction.
+    pub fn new() -> Self {
+        let parents = ParentReconstruction::new();
+        let dag = Dag::new();
+        Reconstruction { parents, dag }
+    }
+
+    fn handle_parents_reconstruction_result(
+        &mut self,
+        reconstruction_result: ReconstructionResult<U>,
+    ) -> ReconstructionResult<U> {
+        let (units, requests) = reconstruction_result.into();
```

Comment:
If `impl<U: Unit> From<ReconstructionResult<U>> for (Vec<ReconstructedUnit<U>>, Vec<Request<U::Hasher>>)` is only used here then I wonder if it's adding any clarity. I would likely prefer to see deconstructing a struct here than making this conversion.

Just a personal preference tho, feel free to ignore and resolve.

Reply:
Yeah, you convinced me, made the fields `pub`, and other changes related to the following comments.